API/ev3_timer.c

```diff
@@ -12,268 +16,552 @@
+    // no element found, removal failed
+    errno = ENOENT;
+    return false;
 }
 
-#endif
+bool tmArrayMirror(CallbackBlock *pDest, CallbackBlock *pSrc)
+{
+    // let's make sure there's enough room to copy the array
+    if (!tmArrayEnsureSize(pDest, pSrc->count))
+        return false;
+
+    // copy trivial fields (but do NOT copy the capacity field)
+    pDest->enabled = pSrc->enabled;
+    pDest->timer_fd = pSrc->timer_fd;
+    pDest->count = pSrc->count;
+
+    // copy contents of the elements array
+    memcpy(pDest->elements,
+           pSrc->elements,
+           pSrc->count * sizeof(TimerCallback));
+    return true;
+}
 
-unsigned long long TimerGetUS()
+bool tmArrayEnsureReserve(CallbackBlock *pBlock, size_t minReserve)
 {
-  struct timeval tv;
-  gettimeofday(&tv, 0);
-  return (((unsigned long long)tv.tv_sec) * 1000000) + tv.tv_usec;
+    return tmArrayEnsureSize(pBlock, pBlock->count + minReserve);
 }
 
-unsigned long long TimerGetMS()
+bool tmArrayEnsureSize(CallbackBlock *pBlock, size_t minSize)
 {
-  struct timeval tv;
-  gettimeofday(&tv, 0);
-  return (((unsigned long long)tv.tv_sec) * 1000) + (tv.tv_usec / 1000);
+    size_t newCapacity = pBlock->capacity;
+    // initial allocation size
+    if (newCapacity == 0)
+        newCapacity = INITIAL_CALLBACKS;
+
+    // expand until satisfied
+    while (newCapacity < minSize)
+        newCapacity *= 2;
+
+    // however skip if no change is necessary
+    if (newCapacity == pBlock->capacity)
+        return true;
+
+    // let's try to realloc the array; abort if unsuccessful
+    TimerCallback *newArray = realloc(pBlock->elements, newCapacity * sizeof(TimerCallback));
+    if (newArray == NULL)
+        return false;
+
+    // successful; write the results
+    pBlock->elements = newArray;
+    pBlock->capacity = newCapacity;
+    return true;
 }
 
-unsigned long long TimerGetCS()
+// ENABLE & DISABLE TIMERS
+
+bool tmEnablePeriod(TimerInterval interval)
 {
-  return TimerGetMS() / 10;
+    // check if we are not done already
+    if (tcb.callbacksFor[interval].enabled)
+        return true;
+
+    // otherwise fill the timer parameter struct (arms the timer)
+    int millis = tmGetIntervalMilliseconds(interval);
+    int seconds = millis / 1000;
+    int nanos = (millis % 1000) * 1000000;
+
+    struct itimerspec params = {
+            .it_value = {.tv_sec = seconds, .tv_nsec = nanos,},
+            .it_interval = {.tv_sec = seconds, .tv_nsec = nanos,}
+    };
+
+    // push the result to kernel
+    RETURN_ON_FAIL(timerfd_settime(tcb.callbacksFor[interval].timer_fd, 0, &params, NULL),
+                   "timerfd_settime() (timer startup) failed", false);
+
+    tcb.callbacksFor[interval].enabled = true;
+    return true;
 }
 
+bool tmDisablePeriod(TimerInterval interval)
+{
+    // check if we are not done already
+    if (!tcb.callbacksFor[interval].enabled)
+        return true;
+
+    // otherwise set all to zero (disarms timer)
+    struct itimerspec params = {

```

Comment:
`= {0};`?

Reply:
Yes, that would have been better.

Reply:
Fixed in 91eb9920a7b8cb38cb7c2ee8eb6e54b906afc90c :+1: